rpmautospec/specparser.py

```diff
@@ -0,0 +1,139 @@
+        split_output = output.split("\n")
+        return split_output[0], split_output[1]
+
+    def query(self, path, specfilename):
+        try:
+            with NamedTemporaryFile(mode="w", prefix="rpmautospec-rpmerr-") as errfd:
+                try:
+                    rpm.setLogFile(errfd)
+                    return self._query(path, specfilename)
+                except Exception as err:  # pylint: disable=broad-exception-caught
+                    with open(errfd.name, "r", errors="replace", encoding="utf-8") as rpmerr_read:
+                        raise SpecParserError(rpmerr_read.read()) from err
+        finally:
+            rpm.setLogFile(sys.stderr)
+            rpm.reloadConfig()
+
+
+class NoRPMHooks(ParserHooks):
+    """Gather Epoch/Version/Release definitions through norpm parser"""
+
+    def __init__(self):
+        self.tags = {}
+
+    def tag_found(self, name, value, _tag_raw):
+        """Gather EclusiveArch, ExcludeArch, BuildArch..."""
+        if name in ["epoch", "version", "release"]:
+            self.tags[name] = value
+
+
+class NoRPMSpecParser(SpecParser):
+    """Use NoRPM to parse spec files"""
+
+    def __init__(self):
+        self.registry = system_macro_registry()
+        self.registry.known_norpm_hacks()
+        self.registry["dist"] = ""
+        name, params = AUTORELEASE_MACRO.split("(")
+        params = params.rstrip(")")
+        self.registry.define(name, (AUTORELEASE_DEFINITION, params))
+
+    def query(self, _path, specfilename):
+        registry = copy.deepcopy(self.registry)
+        hooks = NoRPMHooks()
+        with open(specfilename, "r", encoding="utf8", errors="ignore") as fd:
+            try:
+                specfile_expand(fd.read(), registry, hooks)
+            # TODO: catch just NorpmError (available in norpm 1.7+)
+            except Exception as err:
+                raise SpecParserError(f"parsing error: {err}") from err
+        try:
+            release = hooks.tags["release"]
+            epoch_version = hooks.tags["version"]
+        except KeyError as err:
+            raise SpecParserError('"Version:" or "Release:" not set.') from err
+
+        if "epoch" in hooks.tags:
+            epoch_version = hooks.tags["epoch"] + ":" + epoch_version
+        if "%" in epoch_version:
+            raise SpecParserError(f"unexpanded macro in epoch_version: {epoch_version}")
+        return epoch_version, release
+
+
+class AutoSpecParser(SpecParser):
+    """Use either RPMSpecParser or NoRPMSpecParser depending on the current
+    value in environment variable."""
+
+    def __init__(self):
+        self.cache = {}
+
+    def query(self, path, specfilename):
+        """
+        Return (epoch, overriden_release) tuple)
+        """
+        parser = os.environ.get("RPMAUTOSPEC_SPEC_PARSER", "rpm")
+        if parser not in self.cache:
+            self.cache[parser] = RPMSpecParser() if parser == "rpm" else NoRPMSpecParser()
+        parser = self.cache[parser]
```

Comment:
I would use a functools cache here and not default to `norpm` on unknown values of the env variable.